docs/rfc/0003-collaborative-objects.md

```diff
@@ -0,0 +1,215 @@
+state. A major part of the design of collaborative objects therefore is how we
+store and transmit these messages. Messages in this context are just binary
+blobs which need to be passed to the CRDT implementation in any order.
+
+We add a new entry to the `signed_refs` of a project at
+`refs/namespaces/<project>/remotes/refs/remotes/<peer>/rad/collaborative-objects`.
+This points to a tree with the following structure:
+
+```
+.
+|--<object 1 ID>
+|  |--manifest
+|  |--schema
+|  |  |--<schema change 1>
+|  |  |--<schema change 2>
+|  |--<change 1>
+|  |--<change 2>
+|--<object 2 ID>
+|  ...
+```
+
+We refer to a directory in this tree as an "object directory" and the tuple 
+`(manifest, [schema changes], [changes])` as an "object state".
+
+Objects are created with a unique identifier at creation time. This identifier
+is the name of the directory within the `collaborative_objects` tree where the
+data for this object is kept. Each time we make a change to the object we store
+the new change (which is a binary blob) in the object directory and create a
+new commit. We also store a manifest file which contains the name of the data
+type of this object. This file is never changed and is used to allow
+applications to enumerate collaborative objects of a particular type. Type
+names should be some kind of human readable identifier - similar to XML
+namespaces.
+
+Each object is also created with a schema, this is a JSON schema document which
+is itself encoded as an automerge document within the `schema` directory. This
+allows for changes to the schema (as long as they are backwards and forwards
+compatible) to be distributed as additional changes within the `<object
+ID>/schema` directory.
+
+### Viewing the tracking graph
+
+We need to load the state of objects which multiple people within the
+tracking graph are collaborating on. Given a particular object ID we can load
+all the changes from each remote we have replicated and merge them all to
+obtain the final state of the object. With many remotes this could become
+expensive and there are opportunities to do some indexing ahead of time. A more
+pressing concern is that we want to ensure that the schema is respected, more
+on this later.
+
+
+### Updating objects
+
+To make a change to an object we load the existing messages for an object. The 
+application developer provide us with the binary representation of the change
+to that object. We apply the change and ensure that the new object state still
+matches the object schema. At this point the state of the object may depend on
+many contributions from the tracking graph - not just the ones in our own view
+of the project. This is a natural point at which to compact the document using
+automerge's compacted document format. We bundle all the changes into one
+compacted change file and replace all the changes in our view of the object
+with this compacted change file - then create a new commit, update our signed
+refs, and announce.
```

Comment:
In the Automerge model the computations _are_ the object. In general with CRDTs you are actually passing the history of the whole document around, this is how we're able to do commutative merges.

Reply:
Any references to get more up-to-speed with this concept?

Reply:
Bartosz Sypytkowski (who also has an excellent post about HyParView) has a blog series on CRDTs which starts at https://bartoszsypytkowski.com/the-state-of-a-state-based-crdts/

For more academic work there's an overview of CRDTs (as of 2011, lot's has changed since then) here https://hal.inria.fr/file/index/docid/555588/filename/techreport.pdf

There's also a paper on Automerge specifically at https://arxiv.org/pdf/1608.03960.pdf. The algorithm has changed somewhat but it's still a decent overview.